### PR TITLE
CAMEL-13998 - Kerberos authentication for HDFS connections

### PR DESCRIPTION
**Pull request by marius-cornescu**

### CODE REVIEW

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java [hunk 1]
```diff
@@ -171,33 +191,63 @@ private String getString(Map<String, Object> hdfsSettings, String param, String
...
         return strategies;
     }
 
+    private List<String> getKerberosNamedNodeList(Map<String, Object> hdfsSettings) {
+        kerberosNamedNodes = getString(hdfsSettings, "kerberosNamedNodes", kerberosNamedNodes);
+        return Arrays.stream(kerberosNamedNodes.split(",")).distinct().collect(Collectors.toList());
+    }
+
+
+    Configuration getJAASConfiguration() {
```
Comment (oscerd) [inline components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java]: Is there a specific reason to move the jaasconfiguration here? It was a parameter at component level , now it's not even an endpoint option. It doesn't seem right.

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsFileType.java [hunk 2]
```diff
@@ -54,9 +41,21 @@
 import org.apache.hadoop.io.Text;
 import org.apache.hadoop.io.Writable;
 import org.apache.hadoop.io.WritableComparable;
-import org.apache.hadoop.util.Progressable;
 import org.apache.hadoop.util.ReflectionUtils;
 
+import java.io.ByteArrayOutputStream;
```
Comment (oscerd) [inline components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsFileType.java]: This breaks the code style. Please run the module build with sourcecheck profile enabled and check the warnings

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsInfo.java [hunk 3]
```diff
@@ -16,29 +16,29 @@
...
 
-import java.io.IOException;
-import java.net.URI;
-
+import org.apache.camel.component.hdfs.kerberos.KerberosConfiguration;
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.fs.FileSystem;
 import org.apache.hadoop.fs.Path;
 
+import java.io.IOException;
```
Comment (oscerd) [inline components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsInfo.java]: Codestyle

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsComponent.java [hunk 4]
```diff
@@ -16,18 +16,16 @@
...
-import javax.security.auth.login.Configuration;
-
 import org.apache.camel.Endpoint;
 import org.apache.camel.spi.annotations.Component;
 import org.apache.camel.support.DefaultComponent;
 import org.apache.hadoop.fs.FsUrlStreamHandlerFactory;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.net.URL;
```
Comment (oscerd) [inline components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsComponent.java]: Codestyle

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsProducer.java [hunk 5]
```diff
@@ -116,8 +120,12 @@ protected void doStart() throws Exception {
                 log.debug("Creating IdleCheck task scheduled to run every {} millis", config.getCheckIdleInterval());
                 scheduler.scheduleAtFixedRate(new IdleCheck(idleStrategy), config.getCheckIdleInterval(), config.getCheckIdleInterval(), TimeUnit.MILLISECONDS);
             }
+        } catch (Exception e) {
+            LOG.warn("Failed to start the HDFS producer. Caused by: [{}]", e.getMessage());
+            LOG.trace("", e);
```
Comment (oscerd) [inline components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsProducer.java]: Why logging at trace level?

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/kerberos/HdfsKerberosConfigurationFactory.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package org.apache.camel.component.hdfs.kerberos;
```
Comment (oscerd) [inline components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/kerberos/HdfsKerberosConfigurationFactory.java]: Add Apache license header please

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/kerberos/KerberosConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
+package org.apache.camel.component.hdfs.kerberos;
```
Comment (oscerd) [inline components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/kerberos/KerberosConfiguration.java]: Add Apache license header please

File: components/camel-hdfs/src/test/java/org/apache/camel/component/hdfs/kerberos/HdfsKerberosConfigurationFactoryTest.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package org.apache.camel.component.hdfs.kerberos;
+
```
Comment (oscerd) [inline components/camel-hdfs/src/test/java/org/apache/camel/component/hdfs/kerberos/HdfsKerberosConfigurationFactoryTest.java]: Add Apache license header please

File: components/camel-hdfs/src/test/java/org/apache/camel/component/hdfs/kerberos/KerberosConfigurationTest.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
+package org.apache.camel.component.hdfs.kerberos;
```
Comment (oscerd) [inline components/camel-hdfs/src/test/java/org/apache/camel/component/hdfs/kerberos/KerberosConfigurationTest.java]: Add Apache license header please

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsFileType.java (hunk 2 above)
Comment (marius-cornescu) [reply components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsFileType.java]: Is there a codestype profile for intelij I could use? I've done the formatting with the default setting.

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java (hunk 1 above)
Comment (marius-cornescu) [reply components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java]: Good point, will redesign.

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsFileType.java (hunk 2 above)
Comment (oscerd) [reply components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsFileType.java]: The available stuff is here https://github.com/apache/camel/tree/master/buildingtools/src/main/resources

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java [hunk 10]
```diff
@@ -194,7 +189,7 @@ private String getString(Map<String, Object> hdfsSettings, String param, String
 
         splitStrategy = getString(hdfsSettings, "splitStrategy", kerberosNamedNodes);
 
-        if (Objects.nonNull(splitStrategy)) {
+        if (nonNull(splitStrategy)) {
```
Comment (oscerd) [inline components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java]: Here you can use the ObjectHelper class from camel.

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java [hunk 11]
```diff
@@ -618,7 +585,14 @@ public void setKerberosKeytabLocation(String kerberosKeytabLocation) {
     }
 
     public boolean isKerberosAuthentication() {
-        return Objects.nonNull(kerberosNamedNodes) && Objects.nonNull(kerberosConfigFileLocation) && Objects.nonNull(kerberosUsername) && Objects.nonNull(kerberosKeytabLocation)
-                && !kerberosNamedNodes.isEmpty() && !kerberosConfigFileLocation.isEmpty() && !kerberosUsername.isEmpty() && !kerberosKeytabLocation.isEmpty();
+        return nonNullKerberosProperties() && nonEmptyKerberosProperties();
+    }
+
+    private boolean nonNullKerberosProperties() {
+        return nonNull(kerberosNamedNodes) && nonNull(kerberosConfigFileLocation) && nonNull(kerberosUsername) && nonNull(kerberosKeytabLocation);
```
Comment (oscerd) [inline components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java]: same here

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java (hunk 1 above)
Comment (marius-cornescu) [reply components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java]: Redesigned, please re-check.

**Comment by oscerd** [comment]
Thanks, I'll merge this week

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java [hunk 12]
```diff
@@ -171,33 +186,35 @@ private String getString(Map<String, Object> hdfsSettings, String param, String
...
-                        }
-                        HdfsProducer.SplitStrategyType sst = HdfsProducer.SplitStrategyType.valueOf(tokens[0]);
-                        long ssv = Long.valueOf(tokens[1]);
-                        strategies.add(new HdfsProducer.SplitStrategy(sst, ssv));
-                    }
+
+        splitStrategy = getString(hdfsSettings, "splitStrategy", kerberosNamedNodes);
+
+        if (isNotEmpty(splitStrategy)) {
+            String[] strstrategies = splitStrategy.split(",");
```
Comment (omarsmak) [inline components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java]: nit: do you think this makes sense?
```suggestion
            String[] strStrategies = splitStrategy.split(",");
```

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java [hunk 13]
```diff
@@ -171,33 +186,35 @@ private String getString(Map<String, Object> hdfsSettings, String param, String
...
+            for (String strstrategy : strstrategies) {
```
Comment (omarsmak) [inline components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java]: same here 
```suggestion
            for (String strStrategy : strStrategies) {
```

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java (hunk 12 above)
Comment (marius-cornescu) [reply components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java]: Makes sense.
I'll do another pull request for refactoring and other issues found by sonar.

**Comment by oscerd** [comment]
yes, lets do that as a subsequent PR

**Comment by oscerd** [comment]
I think you need to review the tests. What OS are you on?

File: components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java [hunk 14]
```diff
@@ -171,33 +186,35 @@ private String getString(Map<String, Object> hdfsSettings, String param, String
...
+                long ssv = Long.parseLong(tokens[1]);
+                strategies.add(new HdfsProducer.SplitStrategy(sst, ssv));
             }
         }
         return strategies;
     }
 
+    private List<String> getKerberosNamedNodeList(Map<String, Object> hdfsSettings) {
+        kerberosNamedNodes = getString(hdfsSettings, "kerberosNamedNodes", kerberosNamedNodes);
+        return Arrays.stream(kerberosNamedNodes.split(",")).distinct().collect(Collectors.toList());
```
Comment (oscerd) [inline components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java]: This return null in some tests.

Comment (marius-cornescu) [reply components/camel-hdfs/src/main/java/org/apache/camel/component/hdfs/HdfsConfiguration.java]: Fixed.

**Comment by oscerd** [comment]
Thanks merged on master, fixed a missing import and a test. Looks good.

**Pull request closed without merging**